keps/sig-instrumentation/3077-contextual-logging/README.md

```diff
@@ -0,0 +1,955 @@
+
+- Remove the klog text output format.
+
+## Proposal
+
+Log calls that were already converted to structured logging need to be updated
+to use a logger that gets passed into the function in one of two ways:
+- as explicit parameter
+- attached to a `context.Context`
+
+When a function already accepts a context parameter, then that will be used
+instead of adding a separate parameter. This covers most of client-go and
+avoids another big breaking change for the community.
+
+When a function does not accept a context parameter but needs a context for
+calling some other functions, then a context parameter will be added. As a
+positive side effect, such a change can then also remove several `context.TODO`
+calls (currently over 6000 under `pkg`, `cmd`, `staging` and `test`). An
+explicit logger parameter is suitable for functions which don’t need a context
+and never will.
+
+The rationale for not using both context and an explicit logger parameter is
+the risk that the caller passes an updated logger which it didn't add to the
+context. When the context is then passed down to other functions instead of the
+logger, logging in those functions will not produce the expected result. This
+could be avoided by carefully reviewing code which modifies loggers, but
+designing an API so that such a mistake cannot happen seems safer.
+
+`klog.KObj` and similar helper functions get moved to `k8s.io/klogr`. That package
+will also provide aliases for functions and types from `go-logr/logr`. That way,
+a single import statement will be enough in most files. It will also implement
+loggers for klog text output and unit testing.
+
+When a function retrieves a logger from a context parameter and no logger has
+been set there, it will fall back transparently to logging through a global
+logger instance tracked in `k8s.io/klogr`, i.e. log calls never need to check
+whether they have a valid logger.
+
+### User Stories
+
+#### Story 1
+
+kube-scheduler developer Joan [wants to know which
+pod](https://github.com/kubernetes/kubernetes/issues/91633#issuecomment-675074671)
+and which scheduler predicate log messages are associated with.
+
+When kube-scheduler starts processing a pod, it creates a new logger with
+`logger.WithValue(“pod”, klogr.KObj(pod))` and passes that around. While
+iterating over predicates, another logger gets created with
+`logger.WithName(<predicate name>)` and then is used when invoking that
+predicate.
+
+#### Story 2
+
+Scheduler-plugins developer John wants to increase the verbosity of the
+scheduler while it processes a certain pod ([“per-flow additional
+log”](https://github.com/kubernetes-sigs/scheduler-plugins/pull/289)).
+
+He does that by using `logger.V(0)` as logger for important pods and
+`logger.V(2)` as logger for less important ones. Then when the scheduler’s
+verbosity threshold is `-v=1`, a log message emitted with `V(1).InfoS` through
+the updated logger will be printed for important pods and skipped for less
+important ones.
```

Comment:
Is the new feature here that the log level can be set for a log instance?

Then, if I call `log.V(1)`, the actual level is `1 + log.level`?

Reply:
Correct. It's part of the `logr.Logger` API and concept.

@thockin: this use case here would work better if `logger.V(-1)` (= raising verbosity of log messages) was allowed. As it stands, the code that gets called has to use different levels than it normally would. It would be more natural to use the same levels as in other parts and then for important pods use `logger.V(-1)` instead of `logger`.

Reply:
making the log levels indicated in code non-deterministic seems pretty confusing... if I want to see a particular log line, how do I know what to set verbosity to now? if I set verbosity to a level matching a log line and don't see that line logged, how do I know if a parent call masked the log?

Reply:
> making the log levels indicated in code non-deterministic seems pretty confusing...

This is how logr is designed.

If we implement this use case the other way around (code uses "normal" log levels as described in the current Kubernetes conventions, a caller enables additional log output for important contexts via `logger.V(-1)`), then we get the property that you are looking for: a `logger.V(5).Info` log entry is guaranteed to be emitted for -v=5.

Reply:
> This is how logr is designed.

I understand logr supports a caller being the ultimate decider of how verbosely a library it calls gets to log at. That seems useful to be able to do things like quiet info/debug logs for the etcd client.

That doesn't mean we should use that capability within Kubernetes components to vary log verbosity dynamically for the same log lines. That seems difficult to use and understand.

Reply:
It is specifically what the scheduler developers [want to achieve](https://github.com/kubernetes-sigs/scheduler-plugins/pull/289), though. I don't have a strong opinion on whether it is something that should be done.

If it gets done, I would prefer allowing `logger.V(-1)`.